src/ray/object_manager/ownership_based_object_directory.cc

```diff
@@ -95,101 +98,106 @@ rpc::Address GetOwnerAddressFromObjectInfo(const ObjectInfo &object_info) {
 
 }  // namespace
 
-std::shared_ptr<rpc::CoreWorkerClient> OwnershipBasedObjectDirectory::GetClient(
+std::shared_ptr<rpc::CoreWorkerClientInterface> OwnershipBasedObjectDirectory::GetClient(
     const rpc::Address &owner_address) {
-  WorkerID worker_id = WorkerID::FromBinary(owner_address.worker_id());
-  if (worker_id.IsNil()) {
+  if (WorkerID::FromBinary(owner_address.worker_id()).IsNil()) {
     // If an object does not have owner, return nullptr.
     return nullptr;
   }
-  auto it = worker_rpc_clients_.find(worker_id);
-  if (it == worker_rpc_clients_.end()) {
-    it = worker_rpc_clients_
-             .emplace(worker_id, std::make_shared<rpc::CoreWorkerClient>(
-                                     owner_address, client_call_manager_))
-             .first;
-  }
-  return it->second;
+  return owner_client_pool_->GetOrConnect(owner_address);
 }
 
-ray::Status OwnershipBasedObjectDirectory::ReportObjectAdded(
-    const ObjectID &object_id, const NodeID &node_id, const ObjectInfo &object_info) {
+void OwnershipBasedObjectDirectory::ReportObjectAdded(const ObjectID &object_id,
+                                                      const NodeID &node_id,
+                                                      const ObjectInfo &object_info) {
   const WorkerID &worker_id = object_info.owner_worker_id;
-  rpc::Address owner_address = GetOwnerAddressFromObjectInfo(object_info);
-  std::shared_ptr<rpc::CoreWorkerClient> rpc_client = GetClient(owner_address);
-  if (rpc_client == nullptr) {
+  const auto owner_address = GetOwnerAddressFromObjectInfo(object_info);
+  auto owner_client = GetClient(owner_address);
```

Comment:
will this involve an io?

Reply:
Nope. It can create connection if the client is not cached though.